server/src/main/java/org/elasticsearch/health/components/controller/Controller.java

```diff
@@ -0,0 +1,100 @@
+        String instanceHasMasterSummary = masterNode == null ? INSTANCE_HAS_MASTER_RED_SUMMARY : INSTANCE_HAS_MASTER_GREEN_SUMMARY;
+        GetHealthAction.Indicator instanceHasMaster = new GetHealthAction.Indicator(
+            INSTANCE_HAS_MASTER_NAME,
+            instanceHasMasterStatus,
+            instanceHasMasterSummary,
+            (builder, params) -> {
+                builder.object("coordinating_node", xContentBuilder -> {
+                    builder.field("node_id", coordinatingNode.getId());
+                    builder.field("name", coordinatingNode.getName());
+                });
+                builder.object("master_node", xContentBuilder -> {
+                    if (masterNode != null) {
+                        builder.field("node_id", masterNode.getId());
+                        builder.field("name", masterNode.getName());
+                    } else {
+                        builder.nullField("node_id");
+                        builder.nullField("name");
+                    }
+                });
+                return builder;
+            }
+        );
+
+        var noEligibleMasterNodes = noEligibleMasterNodes(clusterState);
+        var lackOfQuorum = lackOfQuorum(clusterState);
+        final HealthStatus status = HealthStatus.aggregate(
+            Stream.of(instanceHasMaster, noEligibleMasterNodes, lackOfQuorum).map(GetHealthAction.Indicator::getStatus)
+        );
+        return new GetHealthAction.Component(
+            "controller",
+            status,
+            Map.of(INSTANCE_HAS_MASTER_NAME, instanceHasMaster, NO_ELIGIBLE_MASTERS, noEligibleMasterNodes, LACK_OF_QUORUM, lackOfQuorum)
+        );
+    }
+
+    static GetHealthAction.Indicator noEligibleMasterNodes(ClusterState clusterState) {
+        for (DiscoveryNode node : clusterState.nodes()) {
+            if (node.getRoles().contains(DiscoveryNodeRole.MASTER_ROLE)) {
+                return GetHealthAction.Indicator.of(NO_ELIGIBLE_MASTERS, HealthStatus.GREEN, "There are eligible master nodes");
+            }
+        }
+        return GetHealthAction.Indicator.of(NO_ELIGIBLE_MASTERS, HealthStatus.RED, "No eligible master nodes");
+    }
+
+    static GetHealthAction.Indicator lackOfQuorum(ClusterState clusterState) {
+        DiscoveryNodes nodes = clusterState.nodes();
+        if (nodes.getSize() == 1 && nodes.getNodes().valuesIt().next().isMasterNode()) {
+            return GetHealthAction.Indicator.of(LACK_OF_QUORUM, HealthStatus.GREEN, "Only one master node");
+        }
+        long amountOfNodesWithVotingRights = clusterState.nodes()
+            .getNodes()
+            .values()
+            .stream()
+            .filter(
+                e -> e.getRoles().contains(DiscoveryNodeRole.MASTER_ROLE) || e.getRoles().contains(DiscoveryNodeRole.VOTING_ONLY_NODE_ROLE)
+            )
+            .count();
+        return amountOfNodesWithVotingRights >= 3
```

Comment:
A quorum is a majority of the nodes in the voting configuration (see `CoordinationMetadata#VotingConfiguration`). Usually there's 3 nodes in the voting config so 2 would be a quorum, but in other situations we may require more.

Reply:
Thank you! I was trying to figure out whether the number of 3 nodes is a constant or we can pick it dynamically.